src/models/adVAE.jl

```diff
@@ -0,0 +1,187 @@
+struct adVAE
+    encoder
+    generator
+    transformer
+end
+
+function adVAE(;idim::Int=1, zdim::Int=1, hdim::Int=64, activation = "relu", nlayers::Int=3, kwargs...)
+    encoder = Flux.Chain(
+        build_mlp(idim, hdim, hdim, nlayers-1, activation=activation),
+        ConditionalDists.SplitLayer(hdim, [zdim,zdim], [identity,softplus])
+    )
+    transformer = Flux.Chain(
+        build_mlp(zdim*2, hdim, hdim, 2, activation=activation),
+        ConditionalDists.SplitLayer(hdim, [zdim,zdim], [identity,softplus])
+    )
+    generator = build_mlp(zdim, hdim, idim, nlayers, activation=activation, lastlayer="linear")
+    
+    return adVAE(encoder, generator, transformer)
+end
+
+function Conv_adVAE(
+        ;isize::Int=1, 
+        in_ch::Int=1, 
+        zdim::Int=64, 
+        nf::Int=16, 
+        extra_layers::Int=0, 
+        hdim::Int=64,
+        depth::Int=3, 
+        activation="relu", 
+        kwargs...
+        )
+
+    encoder = Flux.Chain(
+        ConvEncoder(isize, in_ch, zdim, nf, extra_layers),
+        x->reshape(x, (1,size(x)[end])), # equivalent to flatten
+        ConditionalDists.SplitLayer(zdim, [zdim, zdim], [identity, softplus])
+    )
+    transfromer = Flux.Chain(
+        build_mlp(zdim*2, hdim, hdim, depth, activation=activation),
+        ConditionalDists.SplitLayer(hdim, [zdim,zdim], [identity,softplus])
+    )
+    generator = Flux.Chain(
+        Flux.Dense(zdim, zdim, activation=activation),
+        x->reshape(x, (1,1,1,size(x)[end])), # inverse to flatten
+        ConvDecoder(isize, 1, in_ch, nf, extra_layers) 
+    )
+
+    return adVAE(encoder, generator, transformer)
+end
+
+Flux.@functor adVAE
+
+function (advae::adVAE)(x)
+    μ, Σ = advae.encoder(x)
+    z = μ + Σ * randn(Float32)
+    xᵣ = advae.generator(z)
+end
+
+kl_divergence(μ, Σ) = 0.5 * sum(1 .+ log.(Σ.^2) - μ.^2  - Σ.^2) # actually it is -kl_divergence
+kl_divergence(μ₁, Σ₁, μ₂, Σ₂) = sum(log.(Σ₂ ./ Σ₁) + (Σ₁.^2 + (μ₁ - μ₂).^2) ./ (2*Σ₂.^2) .- 0.5)
+
+function loss(advae::adVAE, x; γ=1e-3, λ=1e-2, mx=1, mz=1)
+    μ, Σ = advae.encoder(x)
+    μₜ, Σₜ = advae.transformer(cat(μ, Σ, dims=1))
+    
+    z = μ + Σ * randn(Float32)
+    zₜ = μₜ + Σₜ * randn(Float32)
+    
+    xᵣ = advae.generator(z)
+    xₜᵣ = advae.generator(zₜ)
+    
+    μᵣ, Σᵣ = advae.encoder(xᵣ)
+    μₜᵣ, Σₜᵣ = advae.encoder(xₜᵣ)
+
+    # 𝓛 for generator => 𝓛 = 𝓛_z + 𝓛_zₜ 
+    𝓛_z = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μᵣ, Σᵣ)
+    𝓛_zₜ = max(0, (mx - Flux.Losses.mse(xᵣ, xₜᵣ))) + γ * max(0, (mz - kl_divergence(μₜᵣ, Σₜᵣ)))
+    𝓛 = 𝓛_z + 𝓛_zₜ
+    # 𝓛ₜ for transformer 
+    𝓛ₜ = kl_divergence(μ, Σ, μₜ, Σₜ)
+    # 𝓛ₑ for encoder
+    𝓛ₑ = Flux.Losses.mse(x, xᵣ) .+ γ * kl_divergence(μ, Σ) 
```

Comment:
Are you sure that the +- signs are all correct and every loss is written in such for that it is minimized? Because here it seems that the kl_divergence is actually maximized when it is with a - sign.

Reply:
yes, it the paper there is mistake in +/- but i corrected it. mybe i should call `𝓛_kld instead of kl_divergence` because correctrly it is `-kld`